src/picom.c

```diff
@@ -1341,6 +1443,40 @@ static bool redirect_start(session_t *ps) {
 		pixman_region32_init(&ps->damage_ring[i]);
 	}
 
+	ps->frame_pacing = true;
+	if (ps->o.legacy_backends || ps->o.benchmark ||
+	    !ps->backend_data->ops->last_render_time) {
+		// Disable frame pacing if we are using a legacy backend or if we are in
+		// benchmark mode, or if the backend doesn't report render time
+		log_info("Disabling frame pacing.");
+		ps->frame_pacing = false;
+	}
+
+	if (ps->present_exists && ps->frame_pacing) {
+		ps->present_event_id = x_new_id(ps->c);
+		auto select_input = xcb_present_select_input(
+		    ps->c, ps->present_event_id, session_get_target_window(ps),
+		    XCB_PRESENT_EVENT_MASK_COMPLETE_NOTIFY);
+		auto notify_msc =
+		    xcb_present_notify_msc(ps->c, session_get_target_window(ps), 0, 0, 1, 0);
+		set_cant_fail_cookie(ps, select_input);
+		set_cant_fail_cookie(ps, notify_msc);
+		ps->present_event = xcb_register_for_special_xge(
+		    ps->c, &xcb_present_id, ps->present_event_id, NULL);
+
+		// Initialize rendering and frame timing statistics, and frame pacing
+		// states.
+		ps->last_msc_instant = 0;
+		ps->last_msc = 0;
+		ps->last_render = 0;
+		ps->target_msc = 0;
+		rolling_avg_reset(ps->frame_time);
+		rolling_max_reset(ps->render_stats);
+	} else if (ps->frame_pacing) {
+		log_error("Present extension is not supported, frame pacing disabled.");
+		ps->frame_pacing = false;
+	}
```

Comment:
maybe it's better to check for the x present extension in the first `if` that disables the frame pacing? in this case, the `if-else-if` will turn into a simple `if` checking if the frame pacing left enabled after the "disabling" `if`.